File: src/jidelnicek/trip/schemas/trip.py
"""
Trip planning Pydantic schemas for API request/response validation.

This module contains all schemas for trip CRUD operations including:
- Trip creation with meal slot configuration
- Trip participant management
- Trip date management and validation
- Trip list with pagination
- Trip updating and archiving
"""

from datetime import date, datetime
from typing import List, Optional, Dict, Any
from uuid import UUID
from decimal import Decimal

from pydantic import BaseModel, Field, validator, model_validator
from pydantic.config import ConfigDict


class TripParticipantBase(BaseModel):
    """Base schema for trip participants."""
    name: Optional[str] = Field(None, max_length=100, description="Participant name")
    number: Optional[int] = Field(None, ge=1, description="Participant number (if not using name)")
    coefficient: Decimal = Field(
        Decimal("100.00"), 
        ge=Decimal("10"), 
        le=Decimal("300"),
        description="Portion coefficient percentage (100 = 100%)"
    )

    @model_validator(mode='after')
    def validate_name_or_number(self):
        """Ensure either name or number is provided."""
        if not self.name and self.number is None:
            raise ValueError("Either name or number must be provided for participant")
        if self.name and self.number is not None:
            raise ValueError("Provide either name or number, not both")
        return self


class TripParticipantCreate(TripParticipantBase):
    """Schema for creating trip participants."""
    pass


class TripParticipantUpdate(BaseModel):
    """Schema for updating trip participants."""
    name: Optional[str] = Field(None, max_length=100)
    number: Optional[int] = Field(None, ge=1)
    coefficient: Optional[Decimal] = Field(None, ge=Decimal("10"), le=Decimal("300"))


class TripParticipantResponse(TripParticipantBase):
    """Schema for trip participant responses."""
    id: UUID
    trip_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TripBase(BaseModel):
    """Base schema for trips."""
    name: str = Field(..., min_length=1, max_length=100, description="Trip name")
    description: Optional[str] = Field(None, max_length=2000, description="Trip description and notes")
    status: str = Field(
        default="planned",
        description="Trip status: planned, active, completed, cancelled"
    )
    start_date: date = Field(..., description="Trip start date")
    end_date: date = Field(..., description="Trip end date")
    meal_slots: List[str] = Field(
        default=["Breakfast", "Lunch", "Dinner"],
        min_items=1,
        max_items=10,
        description="Configured meal slots for the trip"
    )
    recipe_storage_mode: str = Field(
        default="snapshot",
        description="Recipe storage mode: 'snapshot' or 'track_changes'"
    )
    notes: Optional[str] = Field(None, max_length=2000, description="Trip notes")

    @validator('name')
    def validate_name(cls, v):
        if not v.strip():
            raise ValueError("Trip name cannot be empty")
        return v.strip()
    
    @validator('description')
    def validate_description(cls, v):
        if v is not None:
            v = v.strip()
            if len(v) > 2000:
                raise ValueError("Trip description cannot exceed 2000 characters")
            return v if v else None
        return None
    
    @validator('status')
    def validate_status(cls, v):
        valid_statuses = {'planned', 'active', 'completed', 'cancelled'}
        if v not in valid_statuses:
            raise ValueError(f"Trip status must be one of: {', '.join(valid_statuses)}")
        return v

    @validator('meal_slots')
    def validate_meal_slots(cls, v):
        if not v:
            raise ValueError("At least one meal slot is required")
        
        # Check for duplicates
        if len(v) != len(set(v)):
            raise ValueError("Meal slot names must be unique")
        
        # Validate each slot name
        for slot in v:
            if not slot.strip():
                raise ValueError("Meal slot names cannot be empty")
            if len(slot) > 50:
                raise ValueError("Meal slot names cannot exceed 50 characters")
        
        return [slot.strip() for slot in v]

    @validator('recipe_storage_mode')
    def validate_storage_mode(cls, v):
        allowed_modes = ['snapshot', 'track_changes']
        if v not in allowed_modes:
            raise ValueError(f"Recipe storage mode must be one of: {', '.join(allowed_modes)}")
        return v

    @model_validator(mode='after')
    def validate_dates(self):
        """Ensure end date is not before start date."""
        if self.end_date < self.start_date:
            raise ValueError("End date cannot be before start date")
        
        # Check reasonable trip duration (e.g., max 365 days)
        duration = (self.end_date - self.start_date).days
        if duration > 365:
            raise ValueError("Trip duration cannot exceed 365 days")
        
        return self


class TripCreate(TripBase):
    """Schema for creating trips."""
    participants: List[TripParticipantCreate] = Field(
        default_factory=list,
        max_items=20,
        description="Trip participants (max 20)"
    )

    @validator('participants')
    def validate_participants(cls, v):
        if len(v) > 20:
            raise ValueError("Maximum 20 participants allowed per trip")
        
        # Check for duplicate names
        names = [p.name for p in v if p.name]
        if len(names) != len(set(names)):
            raise ValueError("Participant names must be unique")
        
        # Check for duplicate numbers
        numbers = [p.number for p in v if p.number is not None]
        if len(numbers) != len(set(numbers)):
            raise ValueError("Participant numbers must be unique")
        
        return v


class TripUpdate(BaseModel):
    """Schema for updating trips."""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    meal_slots: Optional[List[str]] = Field(None, min_items=1, max_items=10)
    recipe_storage_mode: Optional[str] = None
    notes: Optional[str] = Field(None, max_length=2000)
    participants: Optional[List[TripParticipantUpdate]] = None

    @validator('name')
    def validate_name(cls, v):
        if v is not None and not v.strip():
            raise ValueError("Trip name cannot be empty")
        return v.strip() if v else v

    @validator('meal_slots')
    def validate_meal_slots(cls, v):
        if v is not None:
            if not v:
                raise ValueError("At least one meal slot is required")
            
            # Check for duplicates
            if len(v) != len(set(v)):
                raise ValueError("Meal slot names must be unique")
            
            # Validate each slot name
            for slot in v:
                if not slot.strip():
                    raise ValueError("Meal slot names cannot be empty")
                if len(slot) > 50:
                    raise ValueError("Meal slot names cannot exceed 50 characters")
            
            return [slot.strip() for slot in v]
        return v

    @validator('recipe_storage_mode')
    def validate_storage_mode(cls, v):
        if v is not None:
            allowed_modes = ['snapshot', 'track_changes']
            if v not in allowed_modes:
                raise ValueError(f"Recipe storage mode must be one of: {', '.join(allowed_modes)}")
        return v

    @model_validator(mode='after')
    def validate_dates(self):
        """Ensure date updates maintain consistency."""
        if self.start_date and self.end_date:
            if self.end_date < self.start_date:
                raise ValueError("End date cannot be before start date")
            
            # Check reasonable trip duration
            duration = (self.end_date - self.start_date).days
            if duration > 365:
                raise ValueError("Trip duration cannot exceed 365 days")
        
        return self


class TripListItem(BaseModel):
    """Schema for trip list items (simplified view)."""
    id: UUID
    name: str
    start_date: date
    end_date: date
    duration_days: int
    participant_count: int
    meal_slot_count: int
    recipe_storage_mode: str
    is_archived: bool
    created_at: datetime
    updated_at: datetime
    has_stove: bool = False
    total_meals_planned: int = 0
    completion_percentage: float = 0.0

    model_config = ConfigDict(from_attributes=True)


class TripDaySummary(BaseModel):
    """Schema for trip day summary."""
    id: UUID
    day_number: int
    date: date
    meals_planned: int
    total_calories: Optional[Decimal]
    total_weight_g: Optional[Decimal]
    has_all_meals: bool
    notes: Optional[str]

    model_config = ConfigDict(from_attributes=True)


class TripResponse(TripBase):
    """Schema for detailed trip responses."""
    id: UUID
    user_id: UUID
    is_archived: bool
    created_at: datetime
    updated_at: datetime
    duration_days: int
    participants: List[TripParticipantResponse] = Field(default_factory=list)
    days: List[TripDaySummary] = Field(default_factory=list)
    has_stove: bool = False
    stove_efficiency: Optional[Decimal] = None
    total_meals_planned: int = 0
    completion_percentage: float = 0.0
    total_calories: Optional[Decimal] = None
    total_weight_g: Optional[Decimal] = None
    total_water_ml: Optional[int] = None
    total_fuel_g: Optional[Decimal] = None

    model_config = ConfigDict(from_attributes=True)


class TripListResponse(BaseModel):
    """Schema for paginated trip list responses."""
    items: List[TripListItem]
    total: int
    page: int
    page_size: int
    total_pages: int
    has_next: bool
    has_prev: bool


class TripSearchFilters(BaseModel):
    """Schema for trip search filters."""
    query: Optional[str] = Field(None, description="Search query for trip name")
    start_date_from: Optional[date] = Field(None, description="Filter trips starting from this date")
    start_date_to: Optional[date] = Field(None, description="Filter trips starting until this date")
    end_date_from: Optional[date] = Field(None, description="Filter trips ending from this date")
    end_date_to: Optional[date] = Field(None, description="Filter trips ending until this date")
    min_duration_days: Optional[int] = Field(None, ge=1, description="Minimum trip duration in days")
    max_duration_days: Optional[int] = Field(None, ge=1, description="Maximum trip duration in days")
    min_participants: Optional[int] = Field(None, ge=1, le=20, description="Minimum number of participants")
    max_participants: Optional[int] = Field(None, ge=1, le=20, description="Maximum number of participants")
    is_archived: Optional[bool] = Field(None, description="Filter by archived status")
    has_stove: Optional[bool] = Field(None, description="Filter by stove configuration")
    recipe_storage_mode: Optional[str] = Field(None, description="Filter by recipe storage mode")
    created_after: Optional[datetime] = Field(None, description="Filter by creation date (after)")
    created_before: Optional[datetime] = Field(None, description="Filter by creation date (before)")

    @validator('recipe_storage_mode')
    def validate_storage_mode(cls, v):
        if v is not None:
            allowed_modes = ['snapshot', 'track_changes']
            if v not in allowed_modes:
                raise ValueError(f"Recipe storage mode must be one of: {', '.join(allowed_modes)}")
        return v

    @model_validator(mode='after')
    def validate_date_ranges(self):
        """Validate date range filters."""
        if self.start_date_from and self.start_date_to:
            if self.start_date_from > self.start_date_to:
                raise ValueError("start_date_from cannot be after start_date_to")
        
        if self.end_date_from and self.end_date_to:
            if self.end_date_from > self.end_date_to:
                raise ValueError("end_date_from cannot be after end_date_to")
        
        if self.min_duration_days and self.max_duration_days:
            if self.min_duration_days > self.max_duration_days:
                raise ValueError("min_duration_days cannot be greater than max_duration_days")
        
        if self.min_participants and self.max_participants:
            if self.min_participants > self.max_participants:
                raise ValueError("min_participants cannot be greater than max_participants")
        
        if self.created_after and self.created_before:
            if self.created_after > self.created_before:
                raise ValueError("created_after cannot be after created_before")
        
        return self


class TripDuplicateRequest(BaseModel):
    """Schema for trip duplication request."""
    new_name: Optional[str] = Field(None, min_length=1, max_length=100, description="New name for duplicated trip")
    include_meals: bool = Field(True, description="Include meal assignments in duplicate")
    include_participants: bool = Field(True, description="Include participants in duplicate")
    new_start_date: Optional[date] = Field(None, description="New start date for duplicated trip")


class TripTemplateRequest(BaseModel):
    """Schema for creating trip template."""
    template_name: str = Field(..., min_length=1, max_length=100, description="Template name")
    is_day_template: bool = Field(False, description="Whether this is a day template or full trip template")


class TripCloneRequest(BaseModel):
    """Schema for cloning an existing trip with configurable options."""
    new_name: str = Field(..., min_length=1, max_length=100, description="Name for the cloned trip")
    start_date: date = Field(..., description="Start date for the cloned trip")
    clone_participants: bool = Field(True, description="Clone participants from the original trip")
    clone_meal_slots: bool = Field(True, description="Clone meal slot configuration from the original trip")
    clone_meal_assignments: bool = Field(True, description="Clone meal assignments from the original trip")
    participant_overrides: Optional[List[TripParticipantCreate]] = Field(
        None, 
        max_items=20,
        description="Override participants (if provided, replaces cloned participants)"
    )
    notes: Optional[str] = Field(None, max_length=2000, description="Notes for the cloned trip")

    @validator('new_name')
    def validate_name(cls, v):
        """Ensure trip name is not empty."""
        if not v.strip():
            raise ValueError("Trip name cannot be empty")
        return v.strip()

    @validator('start_date')
    def validate_start_date(cls, v):
        """Ensure start date is not in the past."""
        if v < date.today():
            raise ValueError("Start date cannot be in the past")
        return v

    @validator('participant_overrides')
    def validate_participant_overrides(cls, v):
        """Validate participant overrides if provided."""
        if v is not None and len(v) > 0:
            # Check for duplicate names
            names = [p.name for p in v if p.name]
            if len(names) != len(set(names)):
                raise ValueError("Participant names must be unique")
            
            # Check for duplicate numbers
            numbers = [p.number for p in v if p.number is not None]
            if len(numbers) != len(set(numbers)):
                raise ValueError("Participant numbers must be unique")
        
        return v


class TripStatsResponse(BaseModel):
    """Schema for trip statistics response."""
    total_trips: int
    active_trips: int
    archived_trips: int
    total_trip_days: int
    average_trip_duration: Optional[float]
    average_participants: Optional[float]
    most_common_meal_slots: List[Dict[str, Any]]
    trips_with_stoves: int
    snapshot_mode_trips: int
    track_changes_mode_trips: int


class ParticipantInfo(BaseModel):
    """Schema for participant info in coefficient calculations."""
    id: UUID
    name: str
    coefficient: float
    effective_portion: float


class MealParticipantSummaryResponse(BaseModel):
    """Schema for meal participant summary response."""
    day_id: UUID
    day_number: int
    date: date
    meal_slot: str
    participant_count: int
    effective_count: Decimal
    participants: List[ParticipantInfo]


class DailyParticipantSummaryResponse(BaseModel):
    """Schema for daily participant summary response."""
    day_id: UUID
    day_number: int
    date: date
    total_participants: int
    meal_summaries: Dict[str, MealParticipantSummaryResponse]
    average_effective_count: Decimal


class TripCoefficientSummaryResponse(BaseModel):
    """Schema for trip coefficient summary response."""
    trip_id: UUID
    total_participants: int
    daily_summaries: List[DailyParticipantSummaryResponse]
    meal_slot_totals: Dict[str, Decimal]
    total_effective_days: Decimal

File: frontend/src/schemas/trip.ts
/**
 * Trip validation schemas
 */
import { z } from 'zod';
import {
  uuidSchema,
  dateSchema,
  dateStringSchema,
  decimalSchema,
  positiveDecimalSchema,
  paginationSchema,
} from './common';
import { participantCreateSchema, participantUpdateSchema, participantResponseSchema } from './participant';
import { recipeUnitSchema } from './recipe';

// Trip status validation
export const tripStatusSchema = z.enum(['planned', 'active', 'completed', 'cancelled'], {
  errorMap: () => ({ message: 'Trip status must be: planned, active, completed, or cancelled' }),
});

// Recipe storage mode validation
export const recipeStorageModeSchema = z.enum(['snapshot', 'track_changes'], {
  errorMap: () => ({ message: 'Recipe storage mode must be: snapshot or track_changes' }),
});

// Trip base schema
export const tripBaseSchema = z
  .object({
    name: z
      .string()
      .min(1, 'Trip name is required')
      .max(100, 'Trip name too long')
      .transform((val) => val.trim())
      .refine((val) => val.length > 0, 'Trip name cannot be empty'),
    description: z
      .string()
      .max(2000, 'Description too long')
      .transform((val) => val?.trim() || null)
      .optional(),
    status: tripStatusSchema.default('planned'),
    start_date: dateStringSchema,
    end_date: dateStringSchema,
    meal_slots: z
      .array(z.string().min(1).max(50))
      .min(1, 'At least one meal slot is required')
      .max(10, 'Maximum 10 meal slots allowed')
      .default(['Breakfast', 'Lunch', 'Dinner'])
      .refine(
        (slots) => {
          // Check for duplicates
          return new Set(slots).size === slots.length;
        },
        'Meal slot names must be unique'
      )
      .transform((slots) => slots.map((s) => s.trim())),
    recipe_storage_mode: recipeStorageModeSchema.default('snapshot'),
    notes: z.string().max(2000, 'Notes too long').optional(),
  })
  .refine(
    (data) => {
      const start = new Date(data.start_date);
      const end = new Date(data.end_date);
      return end >= start;
    },
    {
      message: 'End date cannot be before start date',
      path: ['end_date'],
    }
  )
  .refine(
    (data) => {
      const start = new Date(data.start_date);
      const end = new Date(data.end_date);
      const duration = Math.ceil((end.getTime() - start.getTime()) / (1000 * 60 * 60 * 24));
      return duration <= 365;
    },
    {
      message: 'Trip duration cannot exceed 365 days',
      path: ['end_date'],
    }
  );

// Trip creation schema
export const tripCreateSchema = tripBaseSchema.extend({
  participants: z
    .array(participantCreateSchema)
    .max(20, 'Maximum 20 participants allowed per trip')
    .default([])
    .refine(
      (participants) => {
        // Check for duplicate names
        const names = participants.filter((p) => p.name).map((p) => p.name);
        if (names.length !== new Set(names).size) {
          return false;
        }

        // Check for duplicate numbers
        const numbers = participants.filter((p) => p.number !== undefined).map((p) => p.number);
        if (numbers.length !== new Set(numbers).size) {
          return false;
        }

        return true;
      },
      'Participant names and numbers must be unique'
    ),
});

// Trip update schema
export const tripUpdateSchema = z
  .object({
    name: z
      .string()
      .min(1)
      .max(100)
      .transform((val) => val.trim())
      .refine((val) => val.length > 0, 'Trip name cannot be empty')
      .optional(),
    start_date: dateStringSchema.optional(),
    end_date: dateStringSchema.optional(),
    meal_slots: z
      .array(z.string().min(1).max(50))
      .min(1)
      .max(10)
      .refine((slots) => new Set(slots).size === slots.length, 'Meal slot names must be unique')
      .transform((slots) => slots.map((s) => s.trim()))
      .optional(),
    recipe_storage_mode: recipeStorageModeSchema.optional(),
    notes: z.string().max(2000).optional(),
    participants: z.array(participantUpdateSchema).optional(),
  })
  .refine(
    (data) => {
      if (data.start_date && data.end_date) {
        const start = new Date(data.start_date);
        const end = new Date(data.end_date);
        return end >= start;
      }
      return true;
    },
    {
      message: 'End date cannot be before start date',
      path: ['end_date'],
    }
  )
  .refine(
    (data) => {
      if (data.start_date && data.end_date) {
        const start = new Date(data.start_date);
        const end = new Date(data.end_date);
        const duration = Math.ceil((end.getTime() - start.getTime()) / (1000 * 60 * 60 * 24));
        return duration <= 365;
      }
      return true;
    },
    {
      message: 'Trip duration cannot exceed 365 days',
      path: ['end_date'],
    }
  );

// Trip search filters
export const tripSearchFiltersSchema = z
  .object({
    query: z.string().optional(),
    start_date_from: dateStringSchema.optional(),
    start_date_to: dateStringSchema.optional(),
    end_date_from: dateStringSchema.optional(),
    end_date_to: dateStringSchema.optional(),
    min_duration_days: z.number().int().min(1).optional(),
    max_duration_days: z.number().int().min(1).optional(),
    min_participants: z.number().int().min(1).max(20).optional(),
    max_participants: z.number().int().min(1).max(20).optional(),
    is_archived: z.boolean().optional(),
    has_stove: z.boolean().optional(),
    recipe_storage_mode: recipeStorageModeSchema.optional(),
    created_after: dateSchema.optional(),
    created_before: dateSchema.optional(),
  })
  .refine(
    (data) => {
      if (data.start_date_from && data.start_date_to) {
        return new Date(data.start_date_from) <= new Date(data.start_date_to);
      }
      return true;
    },
    {
      message: 'start_date_from cannot be after start_date_to',
      path: ['start_date_from'],
    }
  )
  .refine(
    (data) => {
      if (data.end_date_from && data.end_date_to) {
        return new Date(data.end_date_from) <= new Date(data.end_date_to);
      }
      return true;
    },
    {
      message: 'end_date_from cannot be after end_date_to',
      path: ['end_date_from'],
    }
  )
  .refine(
    (data) => {
      if (data.min_duration_days && data.max_duration_days) {
        return data.min_duration_days <= data.max_duration_days;
      }
      return true;
    },
    {
      message: 'min_duration_days cannot be greater than max_duration_days',
      path: ['min_duration_days'],
    }
  )
  .refine(
    (data) => {
      if (data.min_participants && data.max_participants) {
        return data.min_participants <= data.max_participants;
      }
      return true;
    },
    {
      message: 'min_participants cannot be greater than max_participants',
      path: ['min_participants'],
    }
  );

// Meal slot schemas
export const mealSlotBaseSchema = z.object({
  meal_type: z.string().min(1).max(50),
  meal_date: dateStringSchema,
  participant_count: z.number().int().min(0).default(0),
  notes: z.string().max(500).optional(),
});

export const mealSlotCreateSchema = mealSlotBaseSchema;

export const mealSlotUpdateSchema = z.object({
  participant_count: z.number().int().min(0).optional(),
  notes: z.string().max(500).optional(),
});

export const mealSlotResponseSchema = mealSlotBaseSchema.extend({
  id: uuidSchema,
  trip_id: uuidSchema,
  day_id: uuidSchema,
  day_number: z.number().int().positive(),
  has_assignment: z.boolean(),
  created_at: dateSchema,
  updated_at: dateSchema,
});

// Meal assignment schemas
export const mealAssignmentBaseSchema = z.object({
  meal_slot_id: uuidSchema,
  recipe_id: uuidSchema,
  participant_overrides: z.array(uuidSchema).default([]),
  notes: z.string().max(500).optional(),
});

export const mealAssignmentCreateSchema = mealAssignmentBaseSchema;

export const mealAssignmentUpdateSchema = z.object({
  recipe_id: uuidSchema.optional(),
  participant_overrides: z.array(uuidSchema).optional(),
  notes: z.string().max(500).optional(),
});

export const mealAssignmentResponseSchema = mealAssignmentBaseSchema.extend({
  id: uuidSchema,
  trip_id: uuidSchema,
  effective_participant_count: decimalSchema,
  recipe_snapshot: z.record(z.any()).optional(),
  created_at: dateSchema,
  updated_at: dateSchema,
});

// Meal assignment bulk operations
export const mealAssignmentBulkCreateSchema = z.object({
  assignments: z.array(mealAssignmentCreateSchema).min(1).max(50),
});

export const mealAssignmentBulkUpdateSchema = z.object({
  updates: z.array(
    z.object({
      id: uuidSchema,
      data: mealAssignmentUpdateSchema,
    })
  ),
});

export const mealAssignmentSwapSchema = z.object({
  assignment_id_1: uuidSchema,
  assignment_id_2: uuidSchema,
});

// Trip day schemas
export const tripDaySummarySchema = z.object({
  id: uuidSchema,
  day_number: z.number().int().positive(),
  date: dateStringSchema,
  meals_planned: z.number().int().nonnegative(),
  total_calories: decimalSchema.optional(),
  total_weight_g: decimalSchema.optional(),
  has_all_meals: z.boolean(),
  notes: z.string().optional(),
});

// Trip response schemas
export const tripListItemSchema = z.object({
  id: uuidSchema,
  name: z.string(),
  start_date: dateStringSchema,
  end_date: dateStringSchema,
  duration_days: z.number().int().positive(),
  participant_count: z.number().int().nonnegative(),
  meal_slot_count: z.number().int().nonnegative(),
  recipe_storage_mode: recipeStorageModeSchema,
  is_archived: z.boolean(),
  created_at: dateSchema,
  updated_at: dateSchema,
  has_stove: z.boolean().default(false),
  total_meals_planned: z.number().int().nonnegative().default(0),
  completion_percentage: z.number().min(0).max(100).default(0),
});

export const tripResponseSchema = tripBaseSchema.extend({
  id: uuidSchema,
  user_id: uuidSchema,
  is_archived: z.boolean(),
  created_at: dateSchema,
  updated_at: dateSchema,
  duration_days: z.number().int().positive(),
  participants: z.array(participantResponseSchema).default([]),
  days: z.array(tripDaySummarySchema).default([]),
  has_stove: z.boolean().default(false),
  stove_efficiency: decimalSchema.optional(),
  total_meals_planned: z.number().int().nonnegative().default(0),
  completion_percentage: z.number().min(0).max(100).default(0),
  total_calories: decimalSchema.optional(),
  total_weight_g: decimalSchema.optional(),
  total_water_ml: z.number().int().nonnegative().optional(),
  total_fuel_g: decimalSchema.optional(),
});

export const tripListResponseSchema = z.object({
  items: z.array(tripListItemSchema),
  total: z.number().int().nonnegative(),
  page: z.number().int().positive(),
  page_size: z.number().int().positive(),
  total_pages: z.number().int().nonnegative(),
  has_next: z.boolean(),
  has_prev: z.boolean(),
});

// Trip action schemas
export const tripDuplicateRequestSchema = z.object({
  new_name: z.string().min(1).max(100).optional(),
  include_meals: z.boolean().default(true),
  include_participants: z.boolean().default(true),
  new_start_date: dateStringSchema.optional(),
});

export const tripTemplateRequestSchema = z.object({
  template_name: z.string().min(1).max(100),
  is_day_template: z.boolean().default(false),
});

export const tripCloneRequestSchema = z
  .object({
    new_name: z
      .string()
      .min(1)
      .max(100)
      .transform((val) => val.trim())
      .refine((val) => val.length > 0, 'Trip name cannot be empty'),
    start_date: dateStringSchema,
    clone_participants: z.boolean().default(true),
    clone_meal_slots: z.boolean().default(true),
    clone_meal_assignments: z.boolean().default(true),
    participant_overrides: z.array(participantCreateSchema).max(20).optional(),
    notes: z.string().max(2000).optional(),
  })
  .refine(
    (data) => {
      const start = new Date(data.start_date);
      const today = new Date();
      today.setHours(0, 0, 0, 0);
      return start >= today;
    },
    {
      message: 'Start date cannot be in the past',
      path: ['start_date'],
    }
  )
  .refine(
    (data) => {
      if (data.participant_overrides && data.participant_overrides.length > 0) {
        // Check for duplicate names
        const names = data.participant_overrides.filter((p) => p.name).map((p) => p.name);
        if (names.length !== new Set(names).size) {
          return false;
        }

        // Check for duplicate numbers
        const numbers = data.participant_overrides
          .filter((p) => p.number !== undefined)
          .map((p) => p.number);
        if (numbers.length !== new Set(numbers).size) {
          return false;
        }
      }
      return true;
    },
    {
      message: 'Participant names and numbers must be unique',
      path: ['participant_overrides'],
    }
  );

// Stove configuration schemas
export const stoveTypeSchema = z.enum(['gas', 'liquid', 'solid', 'electric'], {
  errorMap: () => ({ message: 'Stove type must be: gas, liquid, solid, or electric' }),
});

export const stoveCreateSchema = z.object({
  name: z.string().min(1).max(100),
  type: stoveTypeSchema,
  fuel_type: z.string().min(1).max(50),
  base_fuel_consumption: positiveDecimalSchema,
  altitude_factor: decimalSchema.default('1.0'),
  temperature_factor: decimalSchema.default('1.0'),
  wind_factor: decimalSchema.default('1.0'),
  efficiency_percentage: z
    .number()
    .min(10)
    .max(100)
    .default(70),
  notes: z.string().max(500).optional(),
});

export const stoveUpdateSchema = z.object({
  name: z.string().min(1).max(100).optional(),
  type: stoveTypeSchema.optional(),
  fuel_type: z.string().min(1).max(50).optional(),
  base_fuel_consumption: positiveDecimalSchema.optional(),
  altitude_factor: decimalSchema.optional(),
  temperature_factor: decimalSchema.optional(),
  wind_factor: decimalSchema.optional(),
  efficiency_percentage: z.number().min(10).max(100).optional(),
  notes: z.string().max(500).optional(),
});

export const stoveResponseSchema = stoveCreateSchema.extend({
  id: uuidSchema,
  trip_id: uuidSchema,
  created_at: dateSchema,
  updated_at: dateSchema,
});

// Day plan schemas
export const dayPlanUpdateRequestSchema = z.object({
  notes: z.string().max(1000).optional(),
  participant_overrides: z.record(
    uuidSchema,
    z.object({
      is_present: z.boolean(),
      meal_overrides: z.record(z.string(), z.boolean()).optional(),
    })
  ).optional(),
});

export const mealSlotAssignmentRequestSchema = z.object({
  recipe_id: uuidSchema,
  participant_overrides: z.array(uuidSchema).optional(),
  notes: z.string().max(500).optional(),
});

// Shopping list schemas
export const shoppingItemSchema = z.object({
  ingredient_name: z.string(),
  total_quantity: decimalSchema,
  unit: recipeUnitSchema,
  category: z.string().optional(),
  recipes: z.array(z.string()),
  is_custom: z.boolean().default(false),
});

export const dayShoppingListSchema = z.object({
  day_number: z.number().int().positive(),
  date: dateStringSchema,
  items: z.array(shoppingItemSchema),
  total_weight_g: decimalSchema.optional(),
  total_volume_ml: decimalSchema.optional(),
});

export const tripShoppingListSchema = z.object({
  trip_id: uuidSchema,
  generated_at: dateSchema,
  total_items: z.number().int().nonnegative(),
  total_weight_g: decimalSchema,
  total_volume_ml: decimalSchema,
  items_by_category: z.record(z.string(), z.array(shoppingItemSchema)),
  daily_lists: z.array(dayShoppingListSchema).optional(),
});

// Trip statistics schema
export const tripStatsResponseSchema = z.object({
  total_trips: z.number().int().nonnegative(),
  active_trips: z.number().int().nonnegative(),
  archived_trips: z.number().int().nonnegative(),
  total_trip_days: z.number().int().nonnegative(),
  average_trip_duration: z.number().optional(),
  average_participants: z.number().optional(),
  most_common_meal_slots: z.array(z.record(z.any())),
  trips_with_stoves: z.number().int().nonnegative(),
  snapshot_mode_trips: z.number().int().nonnegative(),
  track_changes_mode_trips: z.number().int().nonnegative(),
});

// Type exports
export type TripStatus = z.infer<typeof tripStatusSchema>;
export type RecipeStorageMode = z.infer<typeof recipeStorageModeSchema>;
export type TripBase = z.infer<typeof tripBaseSchema>;
export type TripCreate = z.infer<typeof tripCreateSchema>;
export type TripUpdate = z.infer<typeof tripUpdateSchema>;
export type TripSearchFilters = z.infer<typeof tripSearchFiltersSchema>;
export type MealSlotBase = z.infer<typeof mealSlotBaseSchema>;
export type MealSlotCreate = z.infer<typeof mealSlotCreateSchema>;
export type MealSlotUpdate = z.infer<typeof mealSlotUpdateSchema>;
export type MealSlotResponse = z.infer<typeof mealSlotResponseSchema>;
export type MealAssignmentBase = z.infer<typeof mealAssignmentBaseSchema>;
export type MealAssignmentCreate = z.infer<typeof mealAssignmentCreateSchema>;
export type MealAssignmentUpdate = z.infer<typeof mealAssignmentUpdateSchema>;
export type MealAssignmentResponse = z.infer<typeof mealAssignmentResponseSchema>;
export type MealAssignmentBulkCreate = z.infer<typeof mealAssignmentBulkCreateSchema>;
export type MealAssignmentBulkUpdate = z.infer<typeof mealAssignmentBulkUpdateSchema>;
export type MealAssignmentSwap = z.infer<typeof mealAssignmentSwapSchema>;
export type TripDaySummary = z.infer<typeof tripDaySummarySchema>;
export type TripListItem = z.infer<typeof tripListItemSchema>;
export type TripResponse = z.infer<typeof tripResponseSchema>;
export type TripListResponse = z.infer<typeof tripListResponseSchema>;
export type TripDuplicateRequest = z.infer<typeof tripDuplicateRequestSchema>;
export type TripTemplateRequest = z.infer<typeof tripTemplateRequestSchema>;
export type TripCloneRequest = z.infer<typeof tripCloneRequestSchema>;
export type StoveType = z.infer<typeof stoveTypeSchema>;
export type StoveCreate = z.infer<typeof stoveCreateSchema>;
export type StoveUpdate = z.infer<typeof stoveUpdateSchema>;
export type StoveResponse = z.infer<typeof stoveResponseSchema>;
export type DayPlanUpdateRequest = z.infer<typeof dayPlanUpdateRequestSchema>;
export type MealSlotAssignmentRequest = z.infer<typeof mealSlotAssignmentRequestSchema>;
export type ShoppingItem = z.infer<typeof shoppingItemSchema>;
export type DayShoppingList = z.infer<typeof dayShoppingListSchema>;
export type TripShoppingList = z.infer<typeof tripShoppingListSchema>;
export type TripStatsResponse = z.infer<typeof tripStatsResponseSchema>;